File: src/main/java/march_2025/RotateImage.java
package march_2025;

public class RotateImage {

    public static void main(String[] args) {
        int[][] matrix = {
                {1, 2, 3},
                {4, 5, 6},
                {7, 8, 9}
        };

        RotateImage rotateImage = new RotateImage();
        rotateImage.rotate(matrix);

        System.out.println();
    }

    public void rotate(int[][] matrix) {

        int L = 0;
        int R = matrix[0].length - 1;

        while (L < R) {
            for (int i = 0; i < R - L; i++) {
                int top = L;
                int bottom = R;

                int temp = matrix[top][L + i];

                matrix[top][L + i] = matrix[bottom - i][L];

                matrix[bottom - i][L] = matrix[bottom][R - i];

                matrix[bottom][R - i] = matrix[top + i][R];

                matrix[top + i][R] = temp;
            }

            L++;
            R--;
        }
    }

}

File: python/rotate_image.py
class Solution(object):
    def rotate(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: None Do not return anything, modify matrix in-place instead.
        """
        l = 0
        r = len(matrix) - 1

        # top, l +i
        # bottom, r -i

        while l < r:
            top = l
            bottom = r
            for i in range(r - l):
                tl = matrix[top][l + i]
                matrix[top][l + i] = matrix[bottom - i][l]
                matrix[bottom - i][l] = matrix[bottom][r - i]
                matrix[bottom][r - i] = matrix[top + i][r]
                matrix[top + i][r] = tl
            l += 1
            r -= 1